pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -3402,6 +3488,454 @@ def SeismicActivityExceptionRule(model, k, t):
+                    - b.v_PumpHead[l1, l2, t1] * mh.p_rhog
+                    + b.v_ValveHead[l1, l2, t1] * mh.p_rhog
+                )
+                return process_constraint(constraint)
+            else:
+                return Constraint.Skip
+
+        mh.NodePressure = Constraint(
+            model.s_LLA,
+            model.s_T,
+            rule=NodePressureRule,
+            doc="Pressure at Node L",
+        )
+
+        def MAOPressureRule(b, l1, t1):
+            constraint = b.v_Pressure[l1, t1] <= b.p_Max_AOP
+            return process_constraint(constraint)
+
+        mh.MAOPressure = Constraint(
+            model.s_L - model.s_P,
+            model.s_T,
+            rule=MAOPressureRule,
+            doc="Max allowable pressure rule",
+        )
+
+        mh.objective = Objective(
+            expr=(
+                sum(
+                    sum(
+                        (mh.v_PumpHead[key, t] * 10 + mh.v_ValveHead[key, t])
+                        for key in model.s_LLA
+                    )
+                    for t in model.s_T
+                )
+            ),
+            sense=minimize,
+            doc="Objective function",
+        )
+
+        opt.solve(mh, tee=True)
+
+    elif model.config.hydraulics == Hydraulics.co_optimize:
+        """
+        For the co-optimize method, first, the strategic model is cloned and then,
+        constraints for computing pressures at every node using
+        Hazen-Williams equation is added to the model. The original objective
+        function of minmizing the cost is modified to include of the cost of
+        operating pumps and valves as a function of the deltaP.
+        """
+        # Clone the model to add hydraulics constraints
+        # The cloned model will be a MINLP formulation. The exisiting objective
+        # must be modified to add operating costs for pumps and valves.
+        mh = model.clone()
+        del mh.objective
+
+        mh.p_HW_material_factor_pipeline = Param(
+            initialize=130,
+            mutable=True,
+            units=pyunits.dimensionless,
+            doc="Pipeline material factor for Hazen Williams equation",
+        )
+        mh.v_PumpHead = Var(
+            model.s_LLA,
+            model.s_T,
+            within=NonNegativeReals,
+            initialize=0,
+            units=pyunits.meter,
+            doc="Pump Head added in the direction of flow, m",
+        )
+        mh.v_ValveHead = Var(
+            model.s_LLA,
+            model.s_T,
+            within=NonNegativeReals,
+            initialize=0,
+            units=pyunits.meter,
+            doc="Valve Head removed in the direction of flow, m",
+        )
+        mh.p_rhog = Param(
+            initialize=9.8 * 1000,
+            units=model.model_units["pressure"] / pyunits.meter,
+            doc="g (m/s2) * density of PW (assumed to be 1000 kg/m3)",
+        )
+        mh.p_Min_AOP = Param(
+            initialize=10 * 6895,
```

Comment:
I see `10` here and `14.7` a few liens below - should they match?

Reply:
The numbers have now been moved to the spreadsheet and should be the same everywhere

Reply:
Resolving.